frontend/controller/actions/group.js

```diff
@@ -897,6 +897,70 @@ export default (sbp('sbp/selectors/register', {
       sbp('okTurtles.events/emit', OPEN_MODAL, 'AddMembers')
     }
   },
+  'gi.actions/group/removeUselessIdentityContracts': function ({ contractID, possiblyUselessContractIDs }) {
+    // NOTE: should remove the identity contracts which we don't need to sync anymore
+    //       for users who don't have any common groups, and any common DMs
+    const rootState = sbp('state/vuex/state')
+    const rootGetters = sbp('state/vuex/getters')
+    const me = rootGetters.ourIdentityContractId
+
+    const removeIdentityContracts = () => {
+      const chatRoomIDs = Object.keys(rootGetters.ourDirectMessages)
+      const groupIDs = Object.keys(rootState.contracts)
+        .filter(gID => rootState.contracts[gID].type === 'gi.contracts/group' && gID !== contractID)
+
+      if (!chatRoomIDs.length && !groupIDs.length) {
+        sbp('chelonia/contract/remove', possiblyUselessContractIDs).catch(e => {
+          console.error(`[gi.actions/group/removeUselessIdentityContracts]: ${e.name} thrown by /remove useless identity contracts:`, e)
+        })
+        return
+      }
+
+      sbp('chelonia/contract/wait', [...chatRoomIDs, ...groupIDs]).then(() => {
+        if (rootGetters.ourIdentityContractId && rootGetters.ourIdentityContractId !== me) {
+          return
+        }
+
+        const pending = Object.entries(sbp('okTurtles.eventQueue/queuedInvocations'))
+          .filter(([q]) => typeof q === 'string')
```

Comment:
This is where the main issue lies. As explained in the comments, you must filter out the irrelevant keys. It's not enough to check whether there are any queued events, but you must check whether there are events only for the keys you're interested in.

Reply:
So, here you'd have:

```js
const waitFor = [...chatRoomIDs, ...groupIDs]

sbp('chelonia/contract/wait', waitFor).then(() => {
  // ...
  const pending = Object.entries(sbp('okTurtles.eventQueue/queuedInvocations'))
    .filter(x => waitFor.includes(x))

  if (pending.length) {
    return removeIdentityContracts()
  }
  // ...
})

Reply:
The comment below about the outer scope mostly referred to this `waitFor` here. I.e., you could refresh the list as the list of groups or chatrooms could've changed between the time `wait` was called and `pending` is checked.